### PR TITLE
[Core] Make instance_of_cause work with ray internal error

### PR DESCRIPTION
**Pull request by rkooo567**
## Why are these changes needed?

Imagine you have a controller actor with a worker actor and have a call to controller actor (i.e., Ray tune).

```
Controller -> Worker

ray.get(controller.get_result.remote())
```

If Worker fails, ray.get will raise an exception `RayTaskError`.

Normally when Ray raises an exception, the exception is a subclass of both RayTaskError and the original exception (returned from RayTaskError.as_instance_of_cause). However, when the Ray's internal exception is raised, it doesn't return the original internal error, but only RayTaskError. 

This behavior is confusing, and for users who'd like to obtain the correct "root error", they need to manually access internal private attribute `RayTaskError.cause`. 

The motivation of the change above ^ was because of the following two issues.
https://github.com/ray-project/ray/issues/5862
https://github.com/ray-project/ray/pull/5870
TL;DR some internal error contains an object id, and we couldn't obtain it. In the master, this is not true anymore, and there's no reason to do this anymore.

After this PR, all the internal error will also be contained from an exception, and user can easily access the raised exception or use `exception.as_instance_of_cause()` to access the original exception properly. 

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Review by rickyyx — APPROVED** [review]
So
- before: it would raise TaskError when it's actually actor error?
- now: it raises the right actor error?

How about other errors?

**Comment by rkooo567** [comment]
Yes. 

Before;
- Children of RayError becomes RayTaskError, and everything else becomes the original error

After:
- everything becomes the original error

**Comment by rkooo567** [comment]
cc @justinvyu
